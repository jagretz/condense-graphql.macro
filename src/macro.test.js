/* eslint-disable no-new-func */
const babel = require("@babel/core");

const babelOptions = {
    plugins: ["macros"],
    /*
    Resolve the macro import relative to this filepath.
    @see [filename](https://babeljs.io/docs/en/options#filename)
    */
    filename: __filename
};

describe("condenseGraphql.macro", () => {
    it("should remove unnecessary whitespace from template literal", () => {
        const transform = `
        const condenseGraphql = require("./condenseGraphql.macro.js");
        const result = condenseGraphql(\`replaces \${amount}    whitespace	characters
        with a single space.\`)
        `;

        /*
        Babel API:
        [transform API](https://babeljs.io/docs/en/babel-core#transform)
        transform [options](https://babeljs.io/docs/en/options)
        */
        const { code } = babel.transform(transform, babelOptions);

        expect(code).toMatchSnapshot();

        const fn = new Function(
            "amount",
            `
            ${code};
            return result;
            `
        );
        const result = fn(5);

        expect(result).toBe("replaces 5 whitespace characters with a single space.");
    });

    it("retains whitespace before, in-between, and after template expressions", () => {
        const transform = `
        const condenseGraphql = require("./condenseGraphql.macro.js");
        const result = condenseGraphql(\`retains space \${before}, \${after} and between \${template} \${expressions}.\`)
        `;

        const { code } = babel.transform(transform, babelOptions);

        expect(code).toMatchSnapshot();

        const fn = new Function(
            "before",
            "after",
            "template",
            "expressions",
            `
            ${code};
            return result;
            `
        );
        const result = fn("before", "after", "template", "expressions");

        expect(result).toBe("retains space before,after and between template expressions.");
    });

    it("retains space between template expressions when separated by a new line", () => {
        const transform = `
        const condenseGraphql = require("./condenseGraphql.macro.js");
        const result = condenseGraphql(\`retains space between \${template}
        \${expressions} when separated by a new line.\`)
        `;

        const { code } = babel.transform(transform, babelOptions);

        expect(code).toMatchSnapshot();

        const fn = new Function(
            "template",
            "expressions",
            `
            ${code};
            return result;
            `
        );
        const result = fn("template", "expressions");

        expect(result).toBe(
            "retains space between template expressions when separated by a new line."
        );
    });

    describe("removes unnecessary whitespace from a graphql query", () => {
        it("containing a single object", () => {
            const transform = `
            const condenseGraphql = require("./condenseGraphql.macro.js");
            const result = condenseGraphql(\`
                query {
                    simpleQuery{
                        options
                    }
                }\`)
            `;

            const { code } = babel.transform(transform, babelOptions);
            expect(code).toMatchSnapshot();

            const fn = new Function(
                `
                ${code};
                return result;
                `
            );
            const result = fn();

            expect(result).toBe("query{simpleQuery{options}}");
        });

        it("containing multiple query params", () => {
            const transform = `
            const condenseGraphql = require("./condenseGraphql.macro.js");
            const result = condenseGraphql(\`
                query {
                    simpleQuery{
                        options
                        date
                        age
                    }
                }\`)
            `;

            const { code } = babel.transform(transform, babelOptions);
            expect(code).toMatchSnapshot();

            const fn = new Function(
                `
                ${code};
                return result;
                `
            );
            const result = fn();

            expect(result).toBe("query{simpleQuery{options date age}}");
        });

        it("containing query variables and nested objects", () => {
            const transform = `
            const condenseGraphql = require("./condenseGraphql.macro.js");
            const result = condenseGraphql(\`
                query queryWithVariables(
                    $id: ID,
                    $age: String,
                    $startDate: Date){
                        products(type: $type) {
                            id
                            code
                            options {
                                timeInMilliseconds
                            }
                        }
                    }\`)
            `;

            const { code } = babel.transform(transform, babelOptions);
            expect(code).toMatchSnapshot();

            const fn = new Function(
                `
                ${code};
                return result;
                `
            );
            const result = fn();

            expect(result).toBe(
                "query queryWithVariables($id:ID,$age:String,$startDate:Date){products(type:$type){id code options{timeInMilliseconds}}}"
            );
        });
    });
});
