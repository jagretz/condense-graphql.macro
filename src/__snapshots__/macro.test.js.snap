// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`condenseGraphql.macro removes unnecessary whitespace from a graphql query containing a single object 1`] = `"const result = \`query{simpleQuery{options}}\`;"`;

exports[`condenseGraphql.macro removes unnecessary whitespace from a graphql query containing multiple query params 1`] = `"const result = \`query{simpleQuery{options date age}}\`;"`;

exports[`condenseGraphql.macro removes unnecessary whitespace from a graphql query containing query variables and nested objects 1`] = `"const result = \`query queryWithVariables($id:ID,$age:String,$startDate:Date){products(type:$type){id code options{timeInMilliseconds}}}\`;"`;

exports[`condenseGraphql.macro retains space between template expressions when separated by a new line 1`] = `"const result = \`retains space between \${template} \${expressions} when separated by a new line.\`;"`;

exports[`condenseGraphql.macro retains whitespace before, in-between, and after template expressions 1`] = `"const result = \`retains space \${before},\${after} and between \${template} \${expressions}.\`;"`;

exports[`condenseGraphql.macro should remove unnecessary whitespace from template literal 1`] = `"const result = \`replaces \${amount} whitespace characters with a single space.\`;"`;
